fairlearn/metrics/metrics_engine.py

```diff
@@ -0,0 +1,103 @@
+# Licensed under the MIT License.
+
+import numpy as np
+from sklearn.metrics import recall_score
+
+from . import DisparityResult, MetricResult
+
+_MESSAGE_NON_BINARY = "Array {0} contains values other than 0 and 1"
+
+
+def true_positive_rate(y_actual, y_predict, group_id, sample_weight=None):
+    return metric_by_groups(recall_score, y_actual, y_predict, group_id, sample_weight)
+
+
+def selection_rate(y_actual, y_predict, group_id, sample_weight=None):
+    if sample_weight is not None:
+        raise NotImplementedError("selection_rate and sample_weight")
+
+    def selection_metric(y_a, y_p, sample_weight):
+        return np.mean(y_p)
+
+    return metric_by_groups(selection_metric, y_actual, y_predict, group_id, sample_weight)
+
+
+def metric_by_groups(metric_function, y_actual, y_predict, group_id, sample_weight=None):
+    _check_binary(y_actual, "y_actual")
+    _check_binary(y_predict, "y_predict")
+    result = MetricResult()
+
+    groups = np.unique(group_id)
+
+    result.metric = metric_function(y_actual, y_predict, sample_weight=sample_weight)
+
+    # The slicing we use requires Numpy arrays
+    y_a = np.array(y_actual)
+    y_p = np.array(y_predict)
+    s_w = None
+    if sample_weight is not None:
+        s_w = np.array(sample_weight)
+    for group in groups:
+        group_indices = [i for i, elem in enumerate(group_id) if elem == group]
+        group_actual = y_a[group_indices]
+        group_predict = y_p[group_indices]
+        group_weight = None
+        if s_w is not None:
+            group_weight = s_w[group_indices]
+        result.group_metric[group] = metric_function(group_actual,
+                                                     group_predict,
+                                                     sample_weight=group_weight)
+
+    return result
+
+
+def make_group_metric(metric_function):
+    def wrapper(y_actual, y_predict, group_id, sample_weight=None):
+        return metric_by_groups(metric_function,
+                                y_actual,
+                                y_predict,
+                                group_id,
+                                sample_weight)
+
+    return wrapper
+
+
+def compute_disparity(group_metric_function,
+                      y_actual, y_predict, group_id,
+                      comparison,
+                      sample_weight=None):
+    metrics = group_metric_function(y_actual, y_predict, group_id, sample_weight)
+
+    result = DisparityResult()
+    result.group_metric = metrics.group_metric
+
+    metric_values = np.array(list(result.group_metric.values()))
+
+    if comparison == 'ratio':
+        result.disparity = 1 - np.min(metric_values) / np.max(metric_values)
+    elif comparison == 'diff':
+        result.disparity = np.max(metric_values) - np.min(metric_values)
+    else:
+        raise ValueError("comparison")
+
+    return result
+
+
+def make_disparity_metric(metric_function, comparison):
+
```

Comment:
The argument names should reflect sklearn convention, so i'd go with **y_true** and **y_pred**; we also decided to switch group_id to one of group_ids or group_data--please coordinate with @romanlutz what you want to do. I think I have a bit of a preference for group_data--it is a bit more generic and still signals that we're using this for "group by" type operation.

Reply:
Changed.